File: mobile/lib/features/operation_financiere/domain/entities/element_maintenance.dart
class ElementMaintenance {
  final int? id;
  final int? catalogueElementId;
  final String? catalogueElementLibelle;
  final String? libelle;

  /// Exemplaires posés. Toujours ≥ 1 : une ligne muette vaut une pièce.
  final int quantite;

  /// TOTAL de la ligne — quantité × prix unitaire, pas le prix d'un
  /// exemplaire. C'est ce champ que somment le coût de l'intervention et la
  /// répartition des dettes.
  final double montant;

  /// Fournisseur de la ligne. Null = le partenaire de l'intervention, cas
  /// courant où un seul prestataire fait tout.
  final int? partenaireId;
  final String? partenaireNom;

  const ElementMaintenance({
    this.id,
    this.catalogueElementId,
    this.catalogueElementLibelle,
    this.libelle,
    this.quantite = 1,
    required this.montant,
    this.partenaireId,
    this.partenaireNom,
  });

  String get effectiveLibelle => catalogueElementLibelle ?? libelle ?? '';

  /// Prix d'un exemplaire. Retrouvé par division plutôt que stocké : le total
  /// est la seule valeur qui fait foi, deux champs finiraient par diverger.
  double get prixUnitaire => quantite > 0 ? montant / quantite : montant;

  /// Libellé tel qu'on le lit dans une liste : « Pneu avant × 4 ». Le nombre
  /// n'apparaît qu'à partir de deux — « × 1 » n'apprendrait rien.
  String get libelleAvecQuantite =>
      quantite > 1 ? '$effectiveLibelle × $quantite' : effectiveLibelle;
}

File: backend/src/main/java/com/tmk/vtcmanager/application/domain/operation/ElementMaintenance.java
package com.tmk.vtcmanager.application.domain.operation;

import com.tmk.vtcmanager.application.domain.partenaire.Partenaire;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.math.RoundingMode;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ElementMaintenance {

    private Long id;
    private CatalogueElementMaintenance catalogueElement; // null si entrée libre
    private String libelle;                               // null si catalogue utilisé

    /**
     * Nombre d'exemplaires posés. Null sur les lignes d'avant la quantité, et
     * sur celles qui n'en parlent pas : {@link #getQuantiteEffective()} les
     * traite comme un exemplaire.
     */
    private Integer quantite;

    /**
     * TOTAL de la ligne — quantité × prix unitaire, pas le prix d'un
     * exemplaire. Tout ce qui somme les éléments (coût de l'intervention,
     * répartition des dettes) lit ce champ.
     */
    private BigDecimal montant;

    /**
     * Qui fournit cette ligne. Null = le partenaire de l'intervention : un seul
     * prestataire fait tout, cas de loin le plus fréquent. Renseigné quand le
     * chantier en mêle plusieurs — le garage répare, un autre vend les pièces.
     */
    private Partenaire partenaire;

    /** Retourne le libellé effectif quelle que soit l'origine. */
    public String getEffectiveLibelle() {
        return catalogueElement != null ? catalogueElement.getLibelle() : libelle;
    }

    /** Quantité posée, ramenée à un exemplaire quand elle n'est pas dite. */
    public int getQuantiteEffective() {
        return quantite != null && quantite > 0 ? quantite : 1;
    }

    /**
     * Prix d'un exemplaire, retrouvé par division : le stocker en plus du
     * total ferait deux vérités à tenir d'accord. Deux décimales, comme le
     * montant lui-même.
     */
    public BigDecimal getPrixUnitaire() {
        if (montant == null) return null;
        return montant.divide(BigDecimal.valueOf(getQuantiteEffective()), 2, RoundingMode.HALF_UP);
    }
}
